pyvista/core/filters/image_data.py

```diff
@@ -1604,3 +1607,281 @@ def _update_and_get_output():
         # Restore active scalars
         self.set_active_scalars(scalars, preference='point')  # type: ignore[attr-defined]
         return output
+
+    def label_connectivity(
+        self,
+        *,
+        scalars: str | None = None,
+        scalar_range: Literal['auto'] | VectorLike[float] | None = 'auto',
+        extraction_mode: Literal['all', 'largest', 'seeded'] = 'all',
+        point_seeds: MatrixLike[float] | VectorLike[float] | _vtk.vtkDataSet | None = None,
+        label_mode: Literal['size', 'constant', 'seeds'] = 'size',
+        constant_value: int | None = None,
+        inplace: bool = False,
+        progress_bar: bool = False,
+    ) -> pyvista.ImageData:
+        """Find and label connected regions in a :class:`~pyvista.ImageData`.
+
+        Only points whose `scalar` value is within the `scalar_range` are considered for
+        connectivity. A 4-connectivity is used for 2D images or a 6-connectivity for 3D
+        images. This filter operates on point-based data. If cell=based data are provided,
+        they are re-meshed to a point-based representation using
+        :func:`~pyvista.ImageDataFilters.cells_to_points` and the output is meshed back
+        to a cell-based representation with :func:`~pyvista.ImageDataFilters.points_to_cells`,
+        effectively filtering based on face connectivity. The connected regions are
+        extracted and labelled according to the strategy defined by `extraction_mode`
+        and `label_mode`, respectively. Unconnected regions are labelled with `0` value.
+
+        .. versionadded:: 0.44.2
+
+        Notes
+        -----
+        This filter implements `vtkImageConnectivityFilter
+        <https://vtk.org/doc/nightly/html/classvtkImageConnectivityFilter.html>`_.
+
+        Using a dissimilar type for `scalars` values and `scalar_range` might produce
+        unexpected results.
```

Comment:
@user27182 @tkoyama010  PyVista could issue a warning when the types are dissimilar. Would it be suitable? There are very few warnings in PyVista code at the moment.

Reply:
It is appropriate to raise an error if the type is different. Would that address the issue?

Reply:
Since vtk is not exiting with an error in this situation but silently accepts to filter, for instance, int cell data with a double range, raising an error would be more restrictive than the original vtk class.

At the same time, accepting silently this situation leads to unexpected rounding.

For instance:
```python
segmented_grid = pv.ImageData(dimensions=(4, 3, 3))
segmented_grid.cell_data["Data"] = [0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0]
connected = segmented_grid.label_connectivity(scalar_range=[0.7, 2])

pl = pv.Plotter()
_ = pl.add_mesh(connected.threshold(0.5), show_edges=True)
pl.show(jupyter_backend="notebook")
```
leads to only one labeled region because (probably) `0.7` is rounded to `0` by vtk which is not the expected behavior from a pythonic point of view.

![image](https://github.com/user-attachments/assets/3749c4c5-0bb3-4023-9eb5-13ca0bec68e9)

Reply:
There are many unexpected results and/or bugs that arise from implicit dtype conversions with vtk. I think this kind of thing should be caught and prevented from happening. One idea is to use numpy's dtype casting rules to do the check and raise an error, e.g. do `scalar_range.astype(scalars.dtype)`

Reply:
I ended up adding an explicit `force_float` parameter with the commit I pushed a few minutes ago.

When `True` (default), it performs a numpy cast as you just described. When `False`, casting is not performed, which saves a bit of resources and works fine in many cases. The docstrings warn users about possible unexpected results.

Reply:
That could work, I think this is fine. Though the main issue was mainly catching a dtype mismatch when a scalar range is specified (of floats) is specified for int inputs, right? So we don't have to _always_ force floats, but just in these specific cases? I think `scalar_range.astype(scalars.dtype)` with `scalars.dtype` instead of `float` would only cast in these specific cases, or maybe even just raise a `TypeError` as suggested by @tkoyama010.

Though like I said this is probably fine as-is. Feel free to mark it as resolved.

Reply:
IMO, the main issue is ensuring that the user gets the expected behavior. For instance, when using a `scalar_range=[0.6, 10]`, one would expect the `0` values in `[0, 0, 1,  4, 2, 0]` to be excluded. This is not the case. In short, we need to cast to floats when the `scalar_range` is not made of whole numbers.

Raising a `TypeError` in this case would serve the purpose but prevent the use of integer data. 

Casting systematically to float is also fine but requires additional resources and is not required in some situations, for instance, `scalar_range=[1, 10]`. Note that vtk expects doubles for `scalar_range,` so it always gets cast to floats.

Reply:
Removed `force_float`. Cast to floats only if the data are integers and the bounds are not whole numbers.